File: projects/nearu/ios/App/App/Plugins/BleBeacon/BleBeaconPlugin.swift
/**
 * BleBeaconPlugin — iOS native implementation using CoreBluetooth + CoreLocation.
 *
 * Scans for iBeacon advertisements matching a given event UUID and reports
 * detected beacons (minorId, rssi, distance) via Capacitor event listeners.
 *
 * Required Info.plist entries:
 *   - NSLocationWhenInUseUsageDescription
 *   - NSLocationAlwaysAndWhenInUseUsageDescription
 *   - NSBluetoothAlwaysUsageDescription
 *   - UIBackgroundModes: bluetooth-central, location
 */

import Foundation
import Capacitor
import CoreLocation
import CoreBluetooth

@objc(BleBeaconPlugin)
public class BleBeaconPlugin: CAPPlugin, CAPBridgedPlugin, CLLocationManagerDelegate, CBCentralManagerDelegate {

    public let identifier = "BleBeaconPlugin"
    public let jsName = "BleBeacon"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startScanning", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopScanning", returnType: CAPPluginReturnPromise),
    ]

    private var locationManager: CLLocationManager?
    private var centralManager: CBCentralManager?
    private var beaconRegion: CLBeaconRegion?
    private var eventUuid: UUID?

    // Path loss model constants (matching ble-utils.ts)
    private let pathLossExponent: Double = 2.0
    private let defaultTxPower: Double = -59.0

    // MARK: - Capacitor Plugin Methods

    @objc func startScanning(_ call: CAPPluginCall) {
        guard let uuidString = call.getString("eventUuid"),
              let uuid = UUID(uuidString: uuidString) else {
            call.reject("Invalid or missing eventUuid")
            return
        }

        self.eventUuid = uuid

        // Initialize CoreBluetooth to check BLE availability
        centralManager = CBCentralManager(delegate: self, queue: nil)

        // Initialize CoreLocation for iBeacon ranging
        locationManager = CLLocationManager()
        locationManager?.delegate = self
        locationManager?.requestWhenInUseAuthorization()
        locationManager?.allowsBackgroundLocationUpdates = true

        // Create beacon region for this event UUID
        let region = CLBeaconRegion(
            uuid: uuid,
            identifier: "nearu.event.\(uuidString)"
        )
        region.notifyOnEntry = true
        region.notifyOnExit = true
        self.beaconRegion = region

        // Start monitoring + ranging
        locationManager?.startMonitoring(for: region)

        // For iOS 13+, use CLBeaconIdentityConstraint
        let constraint = CLBeaconIdentityConstraint(uuid: uuid)
        locationManager?.startRangingBeacons(satisfying: constraint)

        call.resolve()
    }

    @objc func stopScanning(_ call: CAPPluginCall) {
        if let region = beaconRegion {
            locationManager?.stopMonitoring(for: region)
            if let uuid = eventUuid {
                let constraint = CLBeaconIdentityConstraint(uuid: uuid)
                locationManager?.stopRangingBeacons(satisfying: constraint)
            }
        }

        locationManager = nil
        centralManager = nil
        beaconRegion = nil
        eventUuid = nil

        call.resolve()
    }

    // MARK: - CLLocationManagerDelegate (iBeacon Ranging)

    public func locationManager(
        _ manager: CLLocationManager,
        didRange beacons: [CLBeacon],
        satisfying beaconConstraint: CLBeaconIdentityConstraint
    ) {
        for beacon in beacons {
            let rssi = beacon.rssi
            let distance = estimateDistance(rssi: Double(rssi))

            // Only report beacons within ~5 meters
            guard distance <= 5.0 else { continue }

            let data: [String: Any] = [
                "minorId": beacon.minor.intValue,
                "rssi": rssi,
                "estimatedDistance": distance,
                "timestamp": Int(Date().timeIntervalSince1970 * 1000)
            ]

            notifyListeners("beaconDetected", data: data)
        }
    }

    public func locationManager(
        _ manager: CLLocationManager,
        didFailRangingFor beaconConstraint: CLBeaconIdentityConstraint,
        error: Error
    ) {
        notifyListeners("scanError", data: ["message": error.localizedDescription])
    }

    // MARK: - CBCentralManagerDelegate

    public func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOff:
            notifyListeners("scanError", data: ["message": "Bluetooth is powered off"])
        case .unauthorized:
            notifyListeners("scanError", data: ["message": "Bluetooth permission denied"])
        case .unsupported:
            notifyListeners("scanError", data: ["message": "BLE not supported on this device"])
        case .poweredOn:
            break // Ready to scan
        default:
            break
        }
    }

    // MARK: - Distance Estimation

    private func estimateDistance(rssi: Double, txPower: Double = -59.0) -> Double {
        guard rssi < 0 else { return 0.1 }
        let ratio = (txPower - rssi) / (10.0 * pathLossExponent)
        let distance = pow(10.0, ratio)
        return max(0.1, min(100.0, distance))
    }
}

File: projects/nearu/src/plugins/ble/android/BleBeaconPlugin.kt
/**
 * BleBeaconPlugin — Android native implementation using Android BLE API.
 *
 * STUB: This file shows the correct architecture for scanning iBeacon devices
 * on Android via Capacitor. It will NOT compile without Android Studio and the
 * Capacitor Android bridge properly configured.
 *
 * Architecture:
 *   JS (Capacitor) → BleBeaconPlugin (Kotlin) → BluetoothLeScanner
 *
 * iBeacon advertisement data format (parsed manually from raw BLE scan):
 *   Bytes 0-1:   Company ID (0x004C = Apple)
 *   Byte 2:      iBeacon type (0x02)
 *   Byte 3:      Data length (0x15 = 21)
 *   Bytes 4-19:  UUID (16 bytes)
 *   Bytes 20-21: Major (2 bytes, big-endian)
 *   Bytes 22-23: Minor (2 bytes, big-endian)
 *   Byte 24:     TX Power (1 byte, signed)
 *
 * Required AndroidManifest.xml permissions:
 *   - android.permission.BLUETOOTH_SCAN
 *   - android.permission.BLUETOOTH_CONNECT
 *   - android.permission.ACCESS_FINE_LOCATION
 *   - android.permission.ACCESS_COARSE_LOCATION
 */

package com.nearu.plugins.ble

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.nio.ByteBuffer
import java.util.UUID

@CapacitorPlugin(name = "BleBeacon")
class BleBeaconPlugin : Plugin() {

    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bleScanner: BluetoothLeScanner? = null
    private var eventUuid: UUID? = null
    private var isScanning = false

    // Path loss model constants (matching ble-utils.ts)
    private val pathLossExponent = 2.0
    private val defaultTxPower = -59.0

    // iBeacon company identifier (Apple)
    private val IBEACON_COMPANY_ID = 0x004C
    private val IBEACON_TYPE: Byte = 0x02
    private val IBEACON_DATA_LENGTH: Byte = 0x15

    @PluginMethod
    fun startScanning(call: PluginCall) {
        val uuidString = call.getString("eventUuid")
        if (uuidString == null) {
            call.reject("Missing eventUuid")
            return
        }

        try {
            eventUuid = UUID.fromString(uuidString)
        } catch (e: IllegalArgumentException) {
            call.reject("Invalid eventUuid format")
            return
        }

        val bluetoothManager =
            context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
        bluetoothAdapter = bluetoothManager?.adapter

        if (bluetoothAdapter == null || !bluetoothAdapter!!.isEnabled) {
            notifyListeners("scanError", JSObject().put("message", "Bluetooth is not enabled"))
            call.reject("Bluetooth is not enabled")
            return
        }

        bleScanner = bluetoothAdapter?.bluetoothLeScanner
        if (bleScanner == null) {
            call.reject("BLE scanner not available")
            return
        }

        startBleScan()
        call.resolve()
    }

    @PluginMethod
    fun stopScanning(call: PluginCall) {
        stopBleScan()
        call.resolve()
    }

    @SuppressLint("MissingPermission")
    private fun startBleScan() {
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .setReportDelay(0)
            .build()

        // Scan for all BLE devices — we filter iBeacon in the callback
        // because Android ScanFilter cannot match iBeacon manufacturer data partially
        bleScanner?.startScan(emptyList<ScanFilter>(), settings, scanCallback)
        isScanning = true
    }

    @SuppressLint("MissingPermission")
    private fun stopBleScan() {
        if (isScanning) {
            bleScanner?.stopScan(scanCallback)
            isScanning = false
        }
        bleScanner = null
        eventUuid = null
    }

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val scanRecord = result.scanRecord ?: return
            val manufacturerData = scanRecord.getManufacturerSpecificData(IBEACON_COMPANY_ID)
                ?: return

            val parsed = parseIBeaconData(manufacturerData) ?: return

            // Filter by event UUID
            if (parsed.uuid != eventUuid) return

            val rssi = result.rssi
            val distance = estimateDistance(rssi.toDouble(), parsed.txPower.toDouble())

            // Only report beacons within ~5 meters
            if (distance > 5.0) return

            val data = JSObject().apply {
                put("minorId", parsed.minor)
                put("rssi", rssi)
                put("estimatedDistance", distance)
                put("timestamp", System.currentTimeMillis())
            }

            notifyListeners("beaconDetected", data)
        }

        override fun onScanFailed(errorCode: Int) {
            val message = when (errorCode) {
                SCAN_FAILED_ALREADY_STARTED -> "Scan already started"
                SCAN_FAILED_APPLICATION_REGISTRATION_FAILED -> "App registration failed"
                SCAN_FAILED_FEATURE_UNSUPPORTED -> "BLE scan not supported"
                SCAN_FAILED_INTERNAL_ERROR -> "Internal BLE error"
                else -> "Scan failed with code $errorCode"
            }
            notifyListeners("scanError", JSObject().put("message", message))
        }
    }

    /**
     * Parse iBeacon manufacturer-specific data.
     *
     * Expected format after company ID (0x004C) is stripped:
     *   Byte 0:      Type (0x02)
     *   Byte 1:      Length (0x15 = 21)
     *   Bytes 2-17:  UUID (16 bytes)
     *   Bytes 18-19: Major (big-endian)
     *   Bytes 20-21: Minor (big-endian)
     *   Byte 22:     TX Power (signed byte)
     */
    private fun parseIBeaconData(data: ByteArray): IBeaconData? {
        if (data.size < 23) return null
        if (data[0] != IBEACON_TYPE || data[1] != IBEACON_DATA_LENGTH) return null

        val uuidBytes = ByteBuffer.wrap(data, 2, 16)
        val mostSig = uuidBytes.long
        val leastSig = uuidBytes.long
        val uuid = UUID(mostSig, leastSig)

        val major = ((data[18].toInt() and 0xFF) shl 8) or (data[19].toInt() and 0xFF)
        val minor = ((data[20].toInt() and 0xFF) shl 8) or (data[21].toInt() and 0xFF)
        val txPower = data[22].toInt() // signed byte

        return IBeaconData(uuid, major, minor, txPower)
    }

    private fun estimateDistance(rssi: Double, txPower: Double = defaultTxPower): Double {
        if (rssi >= 0) return 0.1
        val ratio = (txPower - rssi) / (10.0 * pathLossExponent)
        val distance = Math.pow(10.0, ratio)
        return maxOf(0.1, minOf(100.0, distance))
    }

    private data class IBeaconData(
        val uuid: UUID,
        val major: Int,
        val minor: Int,
        val txPower: Int,
    )
}
